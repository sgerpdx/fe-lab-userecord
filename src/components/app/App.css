@import url('https://fonts.googleapis.com/css2?family=Press+Start+2P&display=swap');

body {
  background-color: grey;
  color: whitesmoke;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  font-family: 'Press Start 2P', cursive;
}

.controlArea {
  justify-content: center;
  align-items: center;
}

.feedbackArea {
  justify-content: center;
  align-items: center;
}
