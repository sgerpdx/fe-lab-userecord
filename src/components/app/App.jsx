import React from 'react';
import { useState, useEffect } from 'react';
//import { styles } from './App.css';
//import { useRecord } from '../hooks/useRecord';

function App() {
  //const { current, undo, redo, record } = useRecord('#FF0000');
  const [loading, setLoading] = useState(true);
  const [colorRecords, setColorRecords] = useState([]);
  const [counter, setCounter] = useState(0);
  const [editCounter, setEditCounter] = useState(0);
  const [current, setCurrent] = useState('#FF0000');
  //const [undo] = useState(null);
  //const [redo] = useState(null);

  console.log('>>>', colorRecords);
  console.log('//Current:', current);
  console.log('|||counter', counter);
  console.log(':::editCounter', editCounter);

  const record = (value) => {
    //current variable to record & push current into colorRecords:
    setCurrent(value);
    setCounter((counter) => counter + 1);
    //colorRecords.push(current);
    // maybe need to rework this function to fit proper hook syntax:
    setColorRecords((colorRecords) => [...colorRecords, value]);
  };

  const undo = () => {
    const diffCondition = counter - editCounter;
    if (diffCondition > 0) {
      //experiment with switching the order of these two calls:
      setEditCounter((editCounter) => editCounter + 1);
      setCurrent(colorRecords[counter - editCounter - 1]);
    } else console.log('no previous records');
  };

  const redo = () => {
    if (editCounter !== 0) {
      //this part is working, though maybe on a delay:
      setEditCounter((editCounter) => editCounter - 1);
      //this part is not working:
      setCurrent(colorRecords[counter - editCounter + 1]);
    } else return 'no forward records';
  };

  useEffect(() => {
    if (counter === 0) {
      setColorRecords((colorRecords) => [...colorRecords, current]);
    }
    setLoading(false);
  }, [current]);

  // useEffect(() => {
  //   // const lastValue = colorRecords[counter - editCounter];
  //   // console.log('>>>LV', lastValue);
  //   setCurrent(colorRecords[counter]);
  //   setLoading(false);
  // }, [editCounter]);

  // useEffect(() => {
  //   redoRecord(editCounter);
  //   setLoading(false);
  // }, [redo]);

  if (loading) return <h2>Loading...</h2>;

  return (
    <>
      <button onClick={undo}>undo</button>
      <button onClick={redo}>redo</button>
      <input
        type="color"
        value={current}
        onChange={({ target }) => record(target.value)}
      />
      <div
        style={{ backgroundColor: current, width: '10rem', height: '10rem' }}
      ></div>
    </>
  );
}

export default App;

// const record = #{
//   foo: "FOO",
//   bar: "BAR"
// };
// const tuple = #["FOO", "BAR"]
