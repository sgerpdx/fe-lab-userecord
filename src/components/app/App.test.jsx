/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, cleanup, screen, waitFor } from '@testing-library/react';
import App from './App';
import userEvent from '@testing-library/user-event';
//import { rest } from 'msw';
//import { setupServer } from 'msw/node';

describe('App component', () => {
  afterEach(() => cleanup());
  it('renders App', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('behaves as intended', async () => {
    render(<App />);

    const title = await screen.getByText(
      'Select a Color to Change the Square:'
    );
    expect(title).toMatchSnapshot();

    const redoButton = await screen.findByRole('button', {
      name: 'Undo Selection',
    });
    expect(redoButton).toMatchSnapshot();

    const undoButton = await screen.findByRole('button', {
      name: 'Redo Selection',
    });
    expect(undoButton).toMatchSnapshot();

    // userEvent.click(screen.getByText('undo'));
    // expect(
    //   screen
    //     .findByRole('heading', { name: 'User Feedback' })
    //     .toEqual('ðŸ¤– -- no previous records')
    // );

    //per the RTL docs there is no support for <input type="color">
  });
});
