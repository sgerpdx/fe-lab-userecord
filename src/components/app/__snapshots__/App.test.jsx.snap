// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component behaves as intended 1`] = `
<h3>
  Select a Color to Change the Square:
</h3>
`;

exports[`App component behaves as intended 2`] = `
<button
  aria-label="Undo Selection"
>
  undo
</button>
`;

exports[`App component behaves as intended 3`] = `
<button
  aria-label="Redo Selection"
>
  redo
</button>
`;

exports[`App component renders App 1`] = `
<DocumentFragment>
  <section
    class="controlArea"
  >
    <h3>
      Select a Color to Change the Square:
    </h3>
    <hr />
    <button
      aria-label="Undo Selection"
    >
      undo
    </button>
    <button
      aria-label="Redo Selection"
    >
      redo
    </button>
    <input
      style="margin: 10px;"
      type="color"
      value="#FF0000"
    />
    <div
      style="background-color: rgb(255, 0, 0); width: 10rem; height: 10rem; border: 1px solid white; padding: 4px;"
    />
  </section>
  <section
    class="feedbackArea"
  >
    <h4
      aria-label="User Feedback"
    >
      &gt;&gt;&gt;
    </h4>
  </section>
</DocumentFragment>
`;
